pkg/sidecar/sidecar_test.go

```diff
@@ -0,0 +1,139 @@
+	def := n.Configured[0]
+	assert.Equal(t, "default", def.Network, "expected default network to be named `default`")
+}
+
+func verifyConfiguredNetwork(t *testing.T, n *MockNetwork) {
+	assert.Len(t, n.Configured, 2, "expected exactly two networks to be configured")
+	con := n.Configured[1]
+	assert.Equal(t, CONFIGUREDNETWORK, con.Network, "expected configured network name to be passed to network driver")
+	assert.Equal(t, CONFIGUREDBANDWIDTH, con.Default.Bandwidth, "expected requested bandwidth to be set")
+}
+
+// planRequestNetworkConfigure is a testground plan.
+// It requests a change to the network configuration via the sync service.
+func planRequestNetworkConfigure(ctx context.Context, runenv *runtime.RunEnv, hostname string, done chan int, doConfigure bool) (err error) {
+	client, err := sync.NewBoundClient(ctx, runenv)
+	if err != nil {
+		return
+	}
+
+	err = client.WaitNetworkInitialized(ctx, runenv)
+	if err != nil {
+		return
+	}
+
+	if doConfigure {
+		netCfg := sync.NetworkConfig{
+			Network: CONFIGUREDNETWORK,
+			State:   NETWORKDONE,
+			Default: sync.LinkShape{
+				Bandwidth: CONFIGUREDBANDWIDTH,
+			},
+		}
+		topic := sync.NetworkTopic(hostname)
+		client.MustPublishAndWait(ctx, topic, netCfg, NETWORKDONE, 1)
+	}
+
+	time.Sleep(time.Second)
+	close(done)
+	return nil
+}
+
+// subtestSidecarNetworking starts planRequstNetworkConfigure and the real sidecar handler.
+// The sidecar handler is setup to use a mock network so we can capture what would be passed to a
+// real docker or kubernetes network.
+func subtestSidecarNetworking(validator func(*testing.T, *MockNetwork), doConfigure bool) func(*testing.T) {
+	return func(t *testing.T) {
+		hostname, runenv := unique_runenv()
+		nw := NewMockNetwork()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		errsCh := make(chan error)
+		planDone := make(chan int)
+
+		// Create Sidecar instance
+		client, err := sync.NewBoundClient(ctx, runenv)
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+		inst, err := NewInstance(client, runenv, hostname, nw)
+		if err != nil {
+			t.Error(err)
+			t.FailNow()
+		}
+
+		// Start the plan
+		go func() {
+			errsCh <- planRequestNetworkConfigure(ctx, runenv, hostname, planDone, doConfigure)
+		}()
+
+		// Start sidecar handler
+		go func() {
+			errsCh <- handler(ctx, inst)
+		}()
+
+		// Wait until the plan finishes or an error occurs.
+		for {
+			select {
+			case <-planDone:
+				validator(t, nw)
+				return
+			case err := <-errsCh:
+				if err != nil {
+					t.Error(err)
+					t.FailNow()
+				}
+			}
+		}
+	}
+}
+
+func TestSidecarConfiguresNetworking(t *testing.T) {
+	// Don't configure the network, accepting the defaults.
+	t.Run("DefaultNetworkConfiguration", subtestSidecarNetworking(verifyDefaultNetwork, false))
+	// Configure the network to change the bandwidth, etc.
+	t.Run("ConfiguredNetwork", subtestSidecarNetworking(verifyConfiguredNetwork, true))
```

Comment:
Just two test cases where each test case uses entirely different parameters is a bit smelly. In my head, these would be two separate `Test...` methods, where `subtestSidecarNetworking` would do some kind of setup, and the asserts would be part of those methods.

There's way too much indirection here IMO.